### PR TITLE
[bluelink] Implement CCS2 protocol support for EU

### PR DESCRIPTION
**Pull request by florian-h05**

### CODE REVIEW

**Comment by florian-h05** [comment]
CCS2 protocol data read has been confirmed to work in private conversation on the community forum.
Control actions remain to be tested.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java [hunk 1]
```diff
@@ -210,56 +214,98 @@ public boolean getVehicleStatus(final IVehicle vehicle, final boolean forceRefre
...
             addStandardHeaders(request);
             addAuthHeaders(request);
 
-            final BaseResponse<VehicleStatusResponse> response = sendRequest(request, new TypeToken<>() {
-            }, "get vehicle status");
-            final VehicleStatusResponse result = response.result();
-            if (result == null || result.vehicleStatusInfo() == null) {
-                return false;
-            }
+            if (ccs2Protocol) {
```
Comment (marcusb) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java:276]: It would be worth seeing if subclassing would make this cleaner, removing the if/else statements and encapsulating the CCS-specific functionality in a subclass. Alternatively, a shared abstract base class and two subclasses.

Comment (florian-h05) [reply bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java:276]: I’ve already CCS2 data fetching up and running, and it mainly is a different record and some minor changes in the vehicle status method. Wrt to command actions, it’s just another utility method. I think subclassing would separate slightly better, but we have to think through how to do it so we don’t make the handlers more complex. I’ll look into it.

Comment (florian-h05) [reply bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java:276]: The issue is that the API is bound the the account (which is good IMO, we don’t want to log in per vehicle, login is shared between legacy and CCS2), but we have to do different API calls in some places, but not in all.

Comment (florian-h05) [reply bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java:276]: Can you have a look at the BluelinkApiEU class and decide whether the if statements are really an issue?

Comment (marcusb) [reply bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java:276]: I'm ok with leaving as is.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java [hunk 2]
```diff
@@ -210,56 +271,98 @@ public boolean getVehicleStatus(final IVehicle vehicle, final boolean forceRefre
...
+                if (result == null || result.vehicleStatusInfo() == null) {
+                    return false;
+                }
+
+                final VehicleStatusInfo statusInfo = result.vehicleStatusInfo();
+                final var location = statusInfo.vehicleLocation();
+                if (location != null && location.coord() != null) {
+                    final var coord = location.coord();
+                    cb.acceptLocation(new PointType(new DecimalType(coord.lat()), new DecimalType(coord.lon()),
+                            new DecimalType(coord.alt())));
```
Comment (marcusb) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java]: possibly duplicated code that could be factored out

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java [hunk 3]
```diff
@@ -384,13 +524,13 @@ public boolean stopCharging(final IVehicle vehicle) throws BluelinkApiException
...
-     * @param limit
-     * @return
+     * @param vehicle the vehicle to send the request for
+     * @param plugType the EV charging type
+     * @param limit the charge limit
+     * @return true on success
      * @throws BluelinkApiException
      */
-    private boolean sendChargeLimitRequest(final IVehicle vehicle, int plugType, int limit)
+    private boolean sendChargeLimitRequest(final IVehicle vehicle, PlugType plugType, int limit)
```
Comment (marcusb) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java]: for consistency, the params can be made `final`, though I realize that isn't the project convention

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+
+            @Override
+            public boolean doorLock() {
+                var r1 = v.cabin().door().row1();
+                var r2 = v.cabin().door().row2();
+
+                boolean frontLeftLocked = r1.driver() != null && r1.driver().lock() > 0;
+                boolean frontRightLocked = r1.passenger() != null && r1.passenger().lock() > 0;
+                boolean backLeftLocked = r2.left() != null && r2.left().lock() > 0;
+                boolean backRightLocked = r2.right() != null && r2.right().lock() > 0;
```
Comment (marcusb) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java]: the conditional could be factored out, also shared with trunkOpen() and hoodOpen()

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+                PowerSupply ps = v.electronics().powerSupply();
+                return (ps.ignition1() != 0 || ps.ignition3() != 0);
+            }
+
+            @Override
+            public boolean doorLock() {
+                var r1 = v.cabin().door().row1();
+                var r2 = v.cabin().door().row2();
+
+                boolean frontLeftLocked = r1.driver() != null && r1.driver().lock() > 0;
```
Comment (marcusb) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java]: I guess "front right" in the UK - doesn't matter here but I wonder if we have that assumption somewhere else. :)

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/handler/BluelinkVehicleHandler.java [hunk 6]
```diff
@@ -345,10 +345,10 @@ public void acceptStatus(final CommonVehicleStatus data) {
             final var reservChargeInfos = evStatus.reservChargeInfos();
             if (reservChargeInfos != null && reservChargeInfos.targetSocList() != null) {
                 for (final var target : reservChargeInfos.targetSocList()) {
-                    if (target.plugType() == EvStatus.ReserveChargeInfo.PlugType.DC) {
+                    if (target.plugType() == PlugType.DC) {
```
Comment (marcusb) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/handler/BluelinkVehicleHandler.java]:
```suggestion
                    final String channel = switch (target.plugType()) {
                      case DC -> CHANNEL_CHARGE_LIMIT_DC;
                      case AC -> CHANNEL_CHARGE_LIMIT_AC;
                    };
                    updateState(GROUP_CHARGING, channel,
                                new QuantityType<>(target.targetSocLevel(), Units.PERCENT));
```

**Review by marcusb — APPROVED** [review]
some minor comments, but looks good!

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java (hunk 5 above)
Comment (florian-h05) [reply bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java]: UK isn't in the EU anymore.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java (hunk 4 above)
Comment (florian-h05) [reply bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java]: Done :+1:

**Comment by florian-h05** [comment]
As control of CCS2 vehicles has been confirmed to work with the "normal" actions according to private conversation on the forum, this is finally ready for review.
I rebased this as we've switched from 5.2.x to 5.3.x development cycle, however the force push contains no changes to the Bluelink code. All Bluelink changes are addes as new commits.

**Review by marcusb — APPROVED** [review]

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java [hunk 7]
```diff
@@ -188,19 +196,16 @@ public boolean getVehicleStatus(final IVehicle vehicle, final boolean forceRefre
...
             throw new BluelinkApiException("Vehicle ID is missing");
         }
 
-        final @Nullable VehicleStatusData data;
-        if (forceRefresh) {
-            final String url = brandConfig.apiBaseUrl + "/api/v1/spa/vehicles/" + vehicleId + "/status";
+        boolean ccs2Protocol = isCcsProtocol(vehicle);
+
+        final @Nullable CommonVehicleStatus data;
+        if (forceRefresh && !ccs2Protocol) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java:207]: [P1] Preserve forced-refresh semantics for CCS2

With `forceRefresh == true`, CCS2 still enters the cached branch and requests only `/ccs2/carstatus/latest`. Both the scheduled forced refresh and the `forceRefresh()` action therefore keep returning cached state and never wake the vehicle, contrary to the documented behavior. The CCS2 flow needs to trigger `GET /ccs2/carstatus`, wait for the asynchronous report, and then fetch `/latest`; please compose that sequence without blocking the shared scheduler and cover the distinct request sequence in a test.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java [hunk 8]
```diff
@@ -210,56 +215,94 @@ public boolean getVehicleStatus(final IVehicle vehicle, final boolean forceRefre
...
+                cb.acceptOdometer(new QuantityType<>(odometer, KILO(METRE)));
+
+                cb.acceptSmartKeyBatteryWarning(state.electronics().smartKey().batteryWarning() > 0);
+
+                final var lastUpdateTime = result.lastUpdateTime();
+                if (lastUpdateTime != null) {
+                    try {
+                        final Instant instant = Instant.ofEpochMilli(Long.parseLong(lastUpdateTime));
+                        cb.acceptLastUpdateTimestamp(instant);
+                    } catch (final DateTimeParseException e) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java:251]: [P2] Catch the exception this conversion actually throws

`Long.parseLong(lastUpdateTime)` throws `NumberFormatException`, so a malformed external value bypasses this catch and fails the whole status refresh.

```suggestion
                    } catch (final NumberFormatException e) {
```

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+
+    public CommonVehicleStatus toCommonVehicleStatus(IVehicle vehicle) {
+        Vehicle v = this.state != null ? this.state.vehicle() : null;
+        if (v == null)
+            return null;
+
+        return new CommonVehicleStatus() {
+            @Override
+            public boolean engine() {
+                PowerSupply ps = v.electronics().powerSupply();
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java:63]: [P1] Use CCS2 `DrivingReady` for the engine/running channel

The CCS2 vehicle object exposes `DrivingReady`; the ignition bits are separate accessory state and are not equivalent. For example, the upstream `eu_kia_sportage_phev_2027_ccs2.json` fixture has `DrivingReady = 0` and `Ignition3 = 1`, so this reports the vehicle as running when it is not. Other real EV fixtures omit `PowerSupply` entirely, which also makes this dereference fail. Please map `Vehicle.DrivingReady`, use it here, and add tests that distinguish it from the ignition fields.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+                var door = v.cabin().door();
+                if (door == null) {
+                    return false;
+                }
+                for (int val : extract(door.row1(), door.row2(), DoorState::lock)) {
+                    if (val > 0) {
+                        return true;
+                    }
+                }
+                return false;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java:78]: [P1] Invert and aggregate the CCS2 lock values correctly

For CCS2, each `Lock` value is `0` when locked and non-zero when unlocked, and the overall locked channel should only be true when all doors are locked. The current loop implements the exact inverse (true if any door is unlocked); the added fixture has four zero values, so its test should expect a locked vehicle.

```suggestion
                for (int val : extract(door.row1(), door.row2(), DoorState::lock)) {
                    if (val != 0) {
                        return false;
                    }
                }
                return true;
```

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+            }
+
+            @Override
+            public boolean hoodOpen() {
+                return v.body() != null && isOpen(v.body().hood());
+            }
+
+            @Override
+            public boolean airCtrlOn() {
+                return v.cabin().hvac().row1().driver().temperature().value().equals("ON");
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java:113]: [P1] Derive HVAC state from blower speed

`Temperature.Value` is either `"OFF"` or the numeric setpoint; it is not `"ON"`. Consequently this method remains false while climate control is active. CCS2 exposes the operational state as `Cabin.HVAC.Row1.Driver.Blower.SpeedLevel`; please map that field and test both an active numeric-temperature payload and an inactive payload.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+
+            @Override
+            public TemperatureValue airTemp() {
+                // Note: Not available in CCS2
+                return null;
+            }
+
+            @Override
+            public boolean defrost() {
+                return v.body().windshield().front().heat().state() > 0;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java:124]: [P1] Map `Defog` instead of the optional `Heat` block

CCS2 reports front defrost in `Body.Windshield.Front.Defog.State`. Real EV6/EV9 CCS2 fixtures omit `Heat`, so this currently throws during every status update for those payloads. Also, state `2` means off, so `> 0` is not the correct conversion. Please map `Defog` and treat only state `1` as on, with coverage for the missing-`Heat` and state-2 cases.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+                if (seat == null) {
+                    return new SeatHeaterState(0, 0, 0, 0);
+                }
+                int[] states = extract(seat.row1(), seat.row2(), s -> s.climate() != null ? s.climate().state() : 0);
+                return new SeatHeaterState(states[0], states[1], states[2], states[3]);
+            }
+
+            @Override
+            public BatteryStatus battery() {
+                return new BatteryStatus(v.electronics().battery().level());
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java:156]: [P1] Do not publish the auxiliary-battery sentinel as 255%

CCS2 pairs `Electronics.Battery.Level` with `SensorReliability`; a level of `255` with unreliable sensor state means no reading. The upstream `eu_kia_ev9_2024_ccs2_battery_sentinel.json` fixture expects this value to be unavailable, but this code publishes 255%. Please map reliability and return no battery status for the sentinel/unreliable case (the handler already skips a null battery), then add that fixture edge case to the tests.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+                            new EvStatus.ReserveChargeInfo.TargetSOC(PlugType.DC.ordinal(), ci.targetSoC().quick(),
+                                    new DrivingRange(ci.dte().targetSoC().quick(), 1))));
+                }
+
+                List<EvStatus.DrivingDistance> drvDistance = null;
+                if (fuelSystem.dte() != null) {
+                    double totalRange = fuelSystem.dte().total();
+                    DrivingRange range = new DrivingRange(totalRange, 1);
+                    drvDistance = List.of(new EvStatus.DrivingDistance(new EvStatus.DrivingDistance.RangeByFuel(range,
+                            vehicle.isElectric() ? range : null, null)));
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2VehicleStatusResponse.java:193]: [P1] Preserve the distinct PHEV ranges and response unit

`IVehicle.isElectric()` is also true for PHEVs, so this assigns the combined total DTE to the EV-only channel. A real CCS2 PHEV payload reports `DTE.Total = 600` and `DTE.EV = 47`; this code publishes 600 km as the EV range. It also discards `DTE.Unit` by hard-coding unit 1. Please add the EV component to `Dte`, map total versus EV range separately, honor the response unit, and cover a PHEV payload in the tests.

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2ControlRequest.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.bluelink.internal.dto.eu.ccs2;
+
+/**
+ * Control command request for CCU/CCS2 vehicles.
+ *
+ * @author Florian Hotze - Initial contribution
+ */
+public record Ccs2ControlRequest(String command) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/dto/eu/ccs2/Ccs2ControlRequest.java:20]: [P3] Remove the unused CCS2 control DTOs

After the action implementation was switched back to the legacy control endpoints, this class and `Ccs2ControlTokenRequest`, `Ccs2ControlTokenResponse`, and `Ccs2StartClimateRequest` have no references. Keeping these payload types suggests a CCS2 control-token flow exists when it does not; please remove the four dead classes from this PR.

**Review by lsiepel — COMMENTED** [review]
Me and codex looked at this and found these comments to be noteworthy.

**Comment by didiergonze** [comment]
Hi @florian-h05,

I have been testing the BlueLink EU implementation with a Hyundai IONIQ 9.

Hyundai's previous EU authentication no longer works for new logins. Based on the new OneApp/CCI authentication implemented by EVCC, I have added (with help of AI) a Java implementation that:

* uses the Hyundai/Kia username and password through the OneApp/CCI login flow;
* performs RSA password encryption and the CCI/CCS token exchanges;
* retains the legacy refresh-token authentication as a fallback;
* preserves the CCI authentication cookies;
* uses a dedicated HTTP client with a larger request-header buffer;
* includes unit tests and configuration documentation.

I tested it using openHAB 5.1.4 with a locally built 5.2 BlueLink bundle. The account bridge and my IONIQ 9 vehicle are now ONLINE. I have also successfully tested CCS2 status retrieval.

My current working branch is here:

https://github.com/didiergonze/openhab-addons/tree/bluelink-cci-test

This branch was created from the `bluelink-v0.4.0` tag and also contains temporary build commits, so I do not intend to submit it directly as a PR.

Would you prefer that I:

1. create a clean CCI-only PR based on your `bluelink-eu-ccs2` branch, so it can be incorporated into PR #20297; or
2. wait until PR #20297 is merged and then create a separate CCI authentication PR against `main`?

Could you also advise whether this should target only `main`, with a possible later backport to `5.2.x`?

Thank you.

**Comment by florian-h05** [comment]
> wait until PR https://github.com/openhab/openhab-addons/pull/20297 is merged and then create a separate CCI authentication PR against main?

I prefer this approach. This PR is already quite complex and review has started (I just need to find the time to address it), making it bigger will just cause "chaos".

> Could you also advise whether this should target only main, with a possible later backport to 5.2.x?

All development has to target `main`. Merged PRs can be backported to 5.2.x later, usually without requiring further work - just a simple git cherry-pick.

**Comment by lsiepel** [comment]
I think we can pretty quickly merge this once the review comments have been fixed.

**Comment by florian-h05** [comment]
Comments should now all be fixed. I "synced" the data mapping logic with the current code of upstream https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api

**Comment by didiergonze** [comment]
At the same time I was writing my comments, @cd-tronic has created a PR https://github.com/openhab/openhab-addons/pull/21702.  Better two have two solutions than none :-) .  I will thus follow this PR and test it.  Thanks for the reaction.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Eight of the nine findings from the previous review are resolved. The CCS2 data mappings, malformed timestamp handling, PHEV ranges and units, and unused DTO cleanup now match the expected behavior.

The forced-refresh path now uses the correct wake and `/latest` endpoints, but its delayed fetch still executes a blocking request on the common HTTP client's scheduler and the scheduled task is not retained for lifecycle cancellation. That finding remains open.

*This review was AI-assisted.*

File: bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java (hunk 7 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.bluelink/src/main/java/org/openhab/binding/bluelink/internal/api/BluelinkApiEU.java]: The wake → delayed `/latest` sequence is now present, but the earlier concurrency concern remains. The delayed callback runs on `httpClient.getScheduler()` and calls `getVehicleStatus()`, whose request path blocks in `Request.send()`. Because this is the common HTTP client's own scheduler, a slow `/latest` request can occupy infrastructure used for HTTP timeouts; the returned task is also discarded, so disposal or reinitialization cannot cancel it and stale callbacks may update the handler later. Please use a lifecycle-owned timer, dispatch the blocking fetch to an executor, retain and cancel the task, and add the wake → delay → `/latest` regression test.

**Comment by florian-h05** [comment]
Should be solved now @lsiepel :+1:

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM
